File: core/src/ru/myitschool/dcrawler/skills/ForceWave.java
package ru.myitschool.dcrawler.skills;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.ai.AIUtils;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.skills.action.SimpleCombinedAction;
import ru.myitschool.dcrawler.skills.patterns.EntityTargetPattern;
import ru.myitschool.dcrawler.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.skills.targeting.TargetDisplayer;
import ru.myitschool.dcrawler.skills.targeting.TilePos;
import ru.myitschool.dcrawler.tiles.ColorTile;
import ru.myitschool.dcrawler.utils.AdvancedArray;
import ru.myitschool.dcrawler.utils.Utils;

/**
 * Created by Voyager on 06.12.2017.
 */
public class ForceWave extends Skill {

    MathAction rollAction = new DiceAction(20);
    MathAction attackAction = new DiceAction(6);
    int distance = 2;

    public ForceWave(final Entity doer) {
        super(doer);
        addPlayContainer().getEntityPlay().addAction(new SimpleCombinedAction(this, attackAction) {
            @Override
            public void successEffect(Target target, int damage, FloatingDamageMark mark) {
                Utils.pushEntity(getDoer(), target.getEntity(), distance);
                mark.addText("Success");
            }
        });
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new TargetDisplayer() {
            @Override
            public boolean targetCreation(int x, int y, Array<TilePos> array, Skill skill) {
                AdvancedArray<Vector2> raytrace = AIUtils.getCellRaytrace(getDoer().getTileX(), getDoer().getTileY(), x, y, 2);
                raytrace.clip(raytrace.size - 2, raytrace.size - 1);
                Array<Integer> poses = AIUtils.getObstructorIndexes(raytrace, true);
                if (poses.size > 0){
                    raytrace.clip(0, poses.get(0) - 1);
                }
                for (Vector2 vector : raytrace){
                    array.add(new TilePos((int) vector.x, (int) vector.y, new ColorTile(Color.YELLOW, 0.4f, true)));
                }
                return true;
            }
        });
    }

    @Override
    public TargetPattern getPattern() {
        return new EntityTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("force_wave.png");
    }

    @Override
    public String getName() {
        return "Force wave";
    }

    @Override
    public String getDescription() {
        return "You repels your target for " + distance + "cells";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 1;
    }

    @Override
    public int getDistanceMin() {
        return 1;
    }

    @Override
    public int getDistanceMax() {
        return 3;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 0;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_AT_WILL;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_ENTITY;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return true;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/ForceWave.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.ai.AIUtils
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleCombinedAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.EntityTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TilePos
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction
import com.wolg_vlad.dcrawler.ui.tiles.ColorTile
import com.wolg_vlad.dcrawler.utils.Utils

/**
 * Created by Voyager on 06.12.2017.
 */
class ForceWave(doer: Entity) : Skill(doer) {
    var attackAction: MathAction = DiceAction(6)
    var distance = 2

    override val displayers: Array<TargetDisplayer> = arrayOf(object : TargetDisplayer {
        override fun targetCreation(x: Int, y: Int, list: MutableList<TilePos>, skill: Skill): Boolean {
            val raytrace = AIUtils.getCellRaytrace(doer.tileX, doer.tileY, x, y, 2)
            raytrace!!.clip(raytrace.size - 2, raytrace.size - 1)
            val poses = AIUtils.getObstructorIndexes(raytrace, true)
            if (poses!!.size > 0) {
                raytrace.clip(0, poses[0]!! - 1)
            }
            for (vector in raytrace) {
                list.add(TilePos(vector!!.x.toInt(), vector.y.toInt(), ColorTile(Color.YELLOW, 0.4f, true)))
            }
            return true
        }
    })
    override val pattern: TargetPattern = EntityTargetPattern(this)
    override val icon: Texture = Texture("force_wave.png")
    override val name: String = "Force wave"
    override val description: String = "You repels your target for " + distance + "cells"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 1
    override val distanceMin: Int = 1
    override val distanceMax: Int = 3
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 0
    override val type: Int = SKILL_TYPE_AT_WILL
    override val targetType: Int = SKILL_TARGET_TYPE_ENTITY
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = true
    override val isWallTargets: Boolean = false

    init {
        addPlayContainer().entityPlay.addAction(object : SimpleCombinedAction(this, attackAction) {
            public override fun successEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                Utils.pushEntity(doer, target.entity, distance)
                mark.addText("Success")
            }
        })
    }
}